feature: Implement s3 service

sporestudio opened the pull request:
Create new service on charge to put and get object from the s3 bucket. The following features were implemented:
- s3 service object with `get` and `put` methods.
- Integrate s3 service to the api.

jorgerodriguez-palisis committed:
chore: Install boto3 library

Install requirements for use boto3 library that allow us to comunicate
with the s3 bucket.

jorgerodriguez-palisis committed:
feat: Implement s3 service

Basic service to put and get objects into s3 bucket. Should check if
credentials are needed to establish connection with the s3 bucket.

jorgerodriguez-palisis committed:
feat: Implement s3 service into api

Integrate new s3 service with the API. Redis is still using yet to store
files' metadata and index of fragments. As soon as possible we should
implement database store method.

jorgerodriguez-palisis committed:
feat: Integrate s3 service into API

Create singleton to access s3 service and implement different methods
for store and get fragments from s3 bucket.

jorgerodriguez-palisis committed:
fix: Fix lint error

sporestudio committed:
refactor: Set props as private and change key format

- Set bucket and s3 property as private so anybody can use outside the
  object.
- Change format key for a cleaner format.

sporestudio committed:
refactor: Change s3 class name in api

Readapt the api to the new class name of the s3 service.

sporestudio committed:
refactor: Avoid hardcoded names

Pass s3 bucket name as an env variable.

sporestudio committed:
refactor: Pass user id to s3 service

jvicg commented on src/api/services/s3_service.py:
```diff
@@ -0,0 +1,57 @@
+        Note: Ensure that the AWS credentials are set in the environment variables
+        """
+        self._s3 = boto3.client("s3")
+        self._bucket = os.getenv("S3_BUCKET_NAME")
```
Raise an exception if S3_BUCKET_NAME is not defined.
```python
        _bucket = os.getenv("S3_BUCKET_NAME")
        if _bucket is None:
            raise BucketNotDefined("The environmental variable 'S3_BUCKET_NAME' must be defined to initialize an S3 instance.")

        self._bucket = _bucket
```

jvicg commented on src/api/services/s3_service.py:
```diff
@@ -0,0 +1,57 @@
+        self._s3 = boto3.client("s3")
+        self._bucket = os.getenv("S3_BUCKET_NAME")
+
+    def store_fragment(self, user_id: str, file_uuid: str, fragment_index: int, data: bytes) -> None:
```
Use a try-except block to raise a custom exception in case of failure when uploading/downloading bucket.

sporestudio committed:
feature: Terraform dynamic files (#68)

* feat: Add dynamic files generation in main.tf

Generate ansible inventory and ansible configuration in a dynamic way
with terraform local file resource.

* chore: Add necessary terraform global vars

Add variables for user, templates path and credentials path.

* feat: Add ansible cfg template

Add template to generate ansible.cfg in a dynamic way.

* refactor: Modify inventory template

Generate ansible inventory in a dynamic way.

* refactor: Change ansible inventory path

* fix: Correct key path

sporestudio committed:
chore: Add custom exceptions

Create custom exceptions for s3 service to handle errors.

sporestudio committed:
refactor: Handle errors in s3 service

Implement custom exceptions in s3 service and handle errors with the
bucket name.

sporestudio committed:
fix: Fix linter errors
